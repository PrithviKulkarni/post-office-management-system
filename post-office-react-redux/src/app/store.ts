import { configureStore, ThunkAction, Action } from '@reduxjs/toolkit';
import letterReducer from '../features/LetterSlice';
import parcelReducer from '../features/ParcelSlice';
import depotReducer from '../features/DepotSlice';
import driverReducer from '../features/DriverSlice' 
export const store = configureStore({
  reducer: {
    letter: letterReducer,
    parcel: parcelReducer,
    depot: depotReducer,
    driver: driverReducer
  },
});

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>;
