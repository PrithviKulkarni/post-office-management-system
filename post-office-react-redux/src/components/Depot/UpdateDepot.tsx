import React, { useState, SyntheticEvent, useEffect } from 'react'
import { useSelector } from 'react-redux';
import { useParams, useNavigate } from 'react-router-dom';


import { Container, Button, Form, InputGroup } from 'react-bootstrap';


import { Depot, defaultDepot } from '../interfaces/depotinterface';
import { updateADepot, getDepotById } from '../../features/DepotSlice';
import { useAppDispatch } from '../../app/hooks';
import { RootState } from '../../app/store';




const UpdateDepot = () => {
    const navigate = useNavigate();
    const dispatch = useAppDispatch();

    const [city, setCity] = useState<string>('');
    const [errorMessage, setErrorMessage] = useState<string>('')
    const [workers, setWorkers] = useState<number>(0);

    let { depId } = useParams();
    let depotId: string = "0";
    if (!depId) { depId = "0"; }
    depotId = depId;
    console.log({ depotId });

    const depots = useSelector((state: RootState) => getDepotById(state, depotId));

    useEffect(() => {
        if (depots?.city) { setCity(depots.city); }
        if (depots?.workers) { setWorkers(depots.workers); }

    }, [depots])



    const submitHandler = (event: SyntheticEvent) => {
        event.preventDefault();

        const validated = [depId, city, workers].every(Boolean);
        try {
            if (validated) {

                const updDepotData: Depot = defaultDepot;

                if (depId) { updDepotData.id = depId; }
                if (city) { updDepotData.city = city; }
                if (workers) { updDepotData.workers = Number(workers); }


                console.log({ updDepotData });


                dispatch(updateADepot(updDepotData));
                navigate('/depots');

            }
        }
        
        catch (error) {
            setErrorMessage("Depot NOT FOUND.");
            console.log('errorMessage: ', errorMessage);
        }

    }





   

        return (
            <Container className="updateDepot" data-testid="updateDepot">

                <Form onSubmit={submitHandler} data-testid="create-depot-city">


                    <InputGroup className="city">
                        <InputGroup.Text id="basic-addon1">Depot City</InputGroup.Text>
                        <Form.Control
                            type="text"
                            aria-describedby="basic-addon1"
                            aria-label="city"
                            size="sm"
                            data-testid="cityIn"
                            name="city"
                            value={city}
                            onChange={event => setCity(event.target.value)}
                        />
                    </InputGroup>


                   


                    <InputGroup className="workers" data-testid="create-driver-workers">
                        <InputGroup.Text id="basic-addon1">Number of Workers</InputGroup.Text>
                        <Form.Control
                            type="text"
                            aria-describedby="basic-addon1"
                            aria-label="workers"
                            size="sm"
                            data-testid="workersIn"
                            name="workers"
                            value={workers}
                            onChange={event => setWorkers(Number(event.target.value))}
                        />
                    </InputGroup>

                    <Button type="submit" className="button" data-testid="button">Update Depot Details</Button>
                    <p>{errorMessage}</p>
                </Form>
            </Container>
        )
    
}

export default UpdateDepot
