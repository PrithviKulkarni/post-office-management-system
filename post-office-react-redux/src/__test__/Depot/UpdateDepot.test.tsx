import { fireEvent, render, screen } from '@testing-library/react'
import { store } from '../../app/store'
import { Provider } from 'react-redux'
import UpdateDepot from '../../components/Depot/UpdateDepot'
import { BrowserRouter } from 'react-router-dom'

it('renders update depot correctly', () => {
  const { asFragment } = render(
    <Provider store={store}>
      <BrowserRouter>
        <UpdateDepot />
      </BrowserRouter>
    </Provider>
  )
  expect(asFragment()).toMatchSnapshot()
})

it('Verifies the update depot render', () => {
  const { getByTestId } = render(
    <Provider store={store}>
      <BrowserRouter>
        <UpdateDepot />
      </BrowserRouter>
    </Provider>
  )
  const button = getByTestId('button')
  expect(button.textContent).toBe('Update Depot Details')
  const city = getByTestId('cityIn')
  const workers = getByTestId('workersIn')
  fireEvent.change(city, { target: { value: 'test' } })
  fireEvent.change(workers, { target: { value: '27' } })
  fireEvent.click(button)
  expect(button).toBeTruthy()
  expect(screen.getByTestId('updateDepot')).toHaveClass('updateDepot')
})
