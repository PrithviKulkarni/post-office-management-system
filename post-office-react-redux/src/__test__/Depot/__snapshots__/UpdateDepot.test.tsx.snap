// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders update depot correctly 1`] = `
<DocumentFragment>
  <div
    class="updateDepot container"
    data-testid="updateDepot"
  >
    <form
      class=""
      data-testid="create-depot-city"
    >
      <div
        class="city input-group"
      >
        <span
          class="input-group-text"
          id="basic-addon1"
        >
          Depot City
        </span>
        <input
          aria-describedby="basic-addon1"
          aria-label="city"
          class="form-control form-control-sm"
          data-testid="cityIn"
          name="city"
          type="text"
          value=""
        />
      </div>
      <div
        class="workers input-group"
        data-testid="create-driver-workers"
      >
        <span
          class="input-group-text"
          id="basic-addon1"
        >
          Number of Workers
        </span>
        <input
          aria-describedby="basic-addon1"
          aria-label="workers"
          class="form-control form-control-sm"
          data-testid="workersIn"
          name="workers"
          type="text"
          value="0"
        />
      </div>
      <button
        class="button btn btn-primary"
        data-testid="button"
        type="submit"
      >
        Update Depot Details
      </button>
      <p />
    </form>
  </div>
</DocumentFragment>
`;
